sleepq.c

```diff
@@ -0,0 +1,161 @@
+#include <common.h>
+#include <queue.h>
+#include <libkern.h>
+#include <sleepq.h>
+#include <thread.h>
+#include <sched.h>
+#include <critical_section.h>
+
+#define SC_TABLESIZE 256 /* Must be power of 2. */
+#define SC_MASK (SC_TABLESIZE - 1)
+#define SC_SHIFT 8
+#define SC_HASH(wc)                                                            \
+  ((((uintptr_t)(wc) >> SC_SHIFT) ^ (uintptr_t)(wc)) & SC_MASK)
+#define SC_LOOKUP(wc) &sleepq_chains[SC_HASH(wc)]
+
+static sleepq_chain_t sleepq_chains[SC_TABLESIZE];
+
+void sleepq_init() {
+  memset(sleepq_chains, 0, sizeof sleepq_chains);
+  for (int i = 0; i < SC_TABLESIZE; i++)
+    LIST_INIT(&sleepq_chains[i].sc_queues);
+}
+
+sleepq_t *sleepq_lookup(void *wchan) {
+  sleepq_chain_t *sc = SC_LOOKUP(wchan);
+
+  sleepq_t *sq;
+
+  LIST_FOREACH(sq, &sc->sc_queues, sq_entry) {
```

Comment:
This doesn't really scale, does it? True, there is a hash table involved, but since it's size is fixed the complexity is still linear. Obviously, such implementation is nice and general, but rather sub-optimal for mutexes.
Perhaps mutexes could keep a pointer to their sleep queue so that lookups won't be necessary.